In [ ]:
%matplotlib inline
import numpy as np
import pandas as pd
import scipy
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns

Challenge: Feedback analysis
 Estimated Time: 2-3 hours
Now that you've spent some time tinkering with an example Naive Bayes Classifier, it's time to build your own.

We've mentioned that Naive Bayes is particularly good for text classification problems. Before we made a spam filter. 

Now we'll perform a sentiment analysis, classifying whether feedback left on a website is either positive or negative.

Again the UCI Machine Learning database has a nice labeled dataset of sentiment labelled sentences for us to use. This dataset was created for the paper From Group to Individual Labels using Deep Features, Kotzias et. al., KDD 2015.

Pick one of the company data files and build your own classifier. When you're satisfied with its performance (at this point just using the accuracy measure shown in the example), test it on one of the other datasets to see how well these kinds of classifiers translate from one context to another.

Include your model and a brief writeup of your feature engineering and selection process to submit and review with your mentor.

# Building a Classifier
using the Yelp data in this file
Accuracy measure

In [ ]:
# Grab and process the raw data.
#data_path = ("https://github.com/bar10009/Thinkful-Bootcamp/blob/master/yelp_labelled")
#sms_raw = pd.read_csv(data_path, delimiter= '\t', header=None)
#sms_raw.columns = ['spam', 'message']

In [ ]:
df = pd.read_csv("/Users/bradleyrhyins/Desktop/Thinkful-Projects/yelp_labelled.txt", sep=r'\\t'
,header=None)
#df.columns=['review', 'label']

Let’s get some basic information about the data. The .shape method tells us the number of rows and columns in the dataframe.

In [ ]:
df.shape

In [ ]:
df.info()

In [123]:
DF.columns

Index([u'label', u'text', u'text length'], dtype='object')

In [ ]:
df.columns= ["text"]

In [ ]:
df.head()

In [ ]:
label = [np.int(s.split("\t")[1]) for s in df["text"]]

In [ ]:
text = [s.split(".\t")[0] for s in df["text"]]

In [ ]:
DF = pd.DataFrame({"text": text, "label": label})

In [ ]:
DF['text length'] = DF['text'].apply(len)
DF.head()

In [ ]:
print DF.text

In [ ]:
g = sns.FacetGrid(data=DF, col='text')
g.map(plt.hist, 'text', bins=50)

In [125]:
keywords = ['impressive', 'good', 'very good', 'better', 'love', 'loved', 'great']
#Increase the number of keywords
#Trying to predict positive
#Increase number of features

for key in keywords:
    # Note that we add spaces around the key so that we're getting the word,
    # not just pattern matching.
    DF[str(key)] = DF.text.str.contains(
        ' ' + str(key) + ' ',
        case=False
    )

In [126]:
DF.head()

label                                               text  text length  \
0      1                            Wow... Loved this place           23   
1      0                                  Crust is not good           17   
2      0           Not tasty and the texture was just nasty           40   
3      1  Stopped by during the late May bank holiday of...           86   
4      1  The selection on the menu was great and so wer...           58   

   impressive   good  very good  better   love  loved  great  
0       False  False      False   False  False   True  False  
1       False  False      False   False  False  False  False  
2       False  False      False   False  False  False  False  
3       False  False      False   False  False   True  False  
4       False  False      False   False  False  False   True

In [128]:
DF.columns
#features

list_features = DF.columns[2:]
data = DF[list_features]
target = DF["label"]

In [129]:
print DF.columns 

Index([u'label', u'text', u'text length', u'impressive', u'good', u'very good',
       u'better', u'love', u'loved', u'great'],
      dtype='object')


In [130]:
print list_features

Index([u'text length', u'impressive', u'good', u'very good', u'better',
       u'love', u'loved', u'great'],
      dtype='object')


In [131]:
print data

     text length  impressive   good  very good  better   love  loved  great
0             23       False  False      False   False  False   True  False
1             17       False  False      False   False  False  False  False
2             40       False  False      False   False  False  False  False
3             86       False  False      False   False  False   True  False
4             58       False  False      False   False  False  False   True
5             45       False  False      False   False  False  False  False
6             39       False  False      False   False  False  False  False
7            110       False  False      False   False  False  False  False
8             24       False  False      False   False  False  False   True
9             13       False  False      False   False  False  False   True
10            23       False  False      False   False  False  False  False
11            17       False  False      False   False  False  False  False
12          

In [132]:
print target

0      1
1      0
2      0
3      1
4      1
5      0
6      0
7      0
8      1
9      1
10     1
11     0
12     0
13     1
14     0
15     0
16     1
17     0
18     0
19     0
20     0
21     1
22     1
23     1
24     1
25     1
26     0
27     1
28     0
29     0
      ..
970    0
971    0
972    0
973    0
974    0
975    0
976    0
977    0
978    0
979    0
980    0
981    0
982    0
983    0
984    0
985    0
986    0
987    0
988    0
989    0
990    0
991    0
992    0
993    0
994    0
995    0
996    0
997    0
998    0
999    0
Name: label, Length: 1000, dtype: int64


In [133]:
from sklearn.naive_bayes import BernoulliNB

# Instantiate our model and store it in a new variable.
bnb = BernoulliNB()

# Fit our model to the data.
bnb.fit(data, target)

# Classify, storing the result in a new variable.
y_pred = bnb.predict(data)

# Display our results.
print("Number of mislabeled points out of a total {} points : {}".format(
    data.shape[0],
    (target != y_pred).sum()
))

Number of mislabeled points out of a total 1000 points : 416


Out of 1000 points we labeled 416 point not accurate and 584 points accurrate 
The model is 58.4% accurate

#Try to improve using more features 